File: Logging/src/main/java/com/bitmovin/player/samples/logging/EventLogger.java
package com.bitmovin.player.samples.logging;

import com.bitmovin.player.api.event.Event;
import com.bitmovin.player.api.event.EventEmitter;
import com.bitmovin.player.api.event.EventListener;

import java.lang.ref.WeakReference;

/**
 * A generic logger that attaches to all specified {@link Event}s of an {@link EventEmitter} and
 * calls the according log listener depending on the configured {@link LoggerConfig#getLogLevel()}.
 * <p>
 * {@link Level}s are inclusive, meaning that all higher log levels will be logged in addition
 * to the configured log level.
 * E.g. if Warning is configured, events in {@link LoggerConfig#getWarningEvents()} will
 * call the {@link EventLogger#warningListener} listener and events in
 * {@link LoggerConfig#getErrorEvents()} will call the {@link EventLogger#errorListener} function.
 * <p>
 * The attached {@link EventEmitter} is held in a {@link WeakReference} in order to avoid
 * potential memory leaks even if not properly detached.
 */
public class EventLogger<T extends Event> {
    private final LoggerConfig<T> loggerConfig;
    private final EventListener<T> errorListener;
    private final EventListener<T> warningListener;
    private final EventListener<T> infoListener;
    private final EventListener<T> debugListener;
    private WeakReference<EventEmitter<T>> eventEmitterReference;

    enum Level {
        Error, Warning, Info, Debug
    }

    EventLogger(LoggerConfig<T> loggerConfig,
                EventListener<T> errorListener,
                EventListener<T> warningListener,
                EventListener<T> infoListener,
                EventListener<T> debugListener
    ) {
        this.loggerConfig = loggerConfig;
        this.errorListener = errorListener;
        this.warningListener = warningListener;
        this.infoListener = infoListener;
        this.debugListener = debugListener;
    }

    public void attach(EventEmitter<T> eventEmitter) {
        this.eventEmitterReference = new WeakReference<>(eventEmitter);

        for (Class<? extends T> event : loggerConfig.getErrorEvents()) {
            eventEmitter.on(event, errorListener);
        }
        if (loggerConfig.getLogLevel() == Level.Error) return;

        for (Class<? extends T> event : loggerConfig.getWarningEvents()) {
            eventEmitter.on(event, warningListener);
        }
        if (loggerConfig.getLogLevel() == Level.Warning) return;

        for (Class<? extends T> event : loggerConfig.getInfoEvents()) {
            eventEmitter.on(event, infoListener);
        }
        if (loggerConfig.getLogLevel() == Level.Info) return;

        for (Class<? extends T> event : loggerConfig.getDebugEvents()) {
            eventEmitter.on(event, debugListener);
        }
    }

    public void detach() {
        if (eventEmitterReference == null || eventEmitterReference.get() == null) return;

        EventEmitter<T> eventEmitter = this.eventEmitterReference.get();
        eventEmitter.off(errorListener);
        eventEmitter.off(warningListener);
        eventEmitter.off(infoListener);
        eventEmitter.off(debugListener);
    }
}

File: LoggingKotlin/src/main/java/com/bitmovin/player/samples/logging/EventLogger.kt
package com.bitmovin.player.samples.logging

import android.util.Log
import com.bitmovin.player.api.event.Event
import com.bitmovin.player.api.event.EventEmitter
import java.lang.ref.WeakReference
import kotlin.reflect.KClass


/**
 * A generic logger that attaches to all specified [Event]s of an [EventEmitter] and  calls the
 * according log function depending on the configured [LoggerConfig.logLevel].
 *
 * [EventLogger.Level]s are inclusive, meaning that all higher log levels will be logged in addition
 * to the configured log level.
 * E.g. if [EventLogger.Level.Warning] is configured, events in [EventLogger.Level.Warning] will
 * call the [logWarning] function and events in [LoggerConfig.errorEvents] will call the [logError]
 * function.
 *
 * If not configured otherwise, logs will be written to the [android.util.Log] logger with the
 * respective log level.
 *
 * The attached [EventEmitter] is held in a [WeakReference] in order to avoid potential memory
 * leaks even if not properly detached.
 */
class EventLogger<T : Event>(
    private val config: LoggerConfig<T>,
    private val logError: (Event) -> Unit = { Log.e(config.tag, it.toString()) },
    private val logWarning: (Event) -> Unit = { Log.w(config.tag, it.toString()) },
    private val logInfo: (Event) -> Unit = { Log.i(config.tag, it.toString()) },
    private val logDebug: (Event) -> Unit = { Log.d(config.tag, it.toString()) },
) {
    enum class Level { Debug, Info, Warning, Error }

    private var eventEmitterReference: WeakReference<EventEmitter<T>>? = null

    fun attach(eventEmitter: EventEmitter<T>) {
        detach()
        this.eventEmitterReference = WeakReference(eventEmitter)
        when (config.logLevel) {
            Level.Debug -> {
                eventEmitter.onAll(config.errorEvents, logError)
                eventEmitter.onAll(config.warningEvents, logWarning)
                eventEmitter.onAll(config.infoEvents, logInfo)
                eventEmitter.onAll(config.debugEvents, logDebug)
            }
            Level.Info -> {
                eventEmitter.onAll(config.errorEvents, logError)
                eventEmitter.onAll(config.warningEvents, logWarning)
                eventEmitter.onAll(config.infoEvents, logInfo)

            }
            Level.Warning -> {
                eventEmitter.onAll(config.errorEvents, logError)
                eventEmitter.onAll(config.warningEvents, logWarning)

            }
            Level.Error -> {
                eventEmitter.onAll(config.errorEvents, logError)
            }
        }
    }

    fun detach() {
        val eventEmitter = eventEmitterReference?.get() ?: return
        with(eventEmitter) {
            off(logError)
            off(logWarning)
            off(logInfo)
            off(logDebug)
        }
    }

    private fun EventEmitter<T>.onAll(
        events: List<KClass<out Event>>, block: (Event) -> Unit
    ) = events.filterIsInstance<KClass<T>>().forEach { on(it, block) }
}

